src/crypto.rs

```diff
@@ -1,158 +1,240 @@
-use block_modes::BlockMode;
-use cipher::generic_array::typenum::{U16, U32};
-use cipher::generic_array::{ArrayLength, GenericArray};
-use cipher::NewBlockCipher;
+use aes::cipher::block_padding::Pkcs7;
+use aes::cipher::{BlockDecryptMut, BlockEncryptMut, KeyIvInit};
+use cipher::generic_array::GenericArray;
 use rand::{thread_rng, RngCore};
 
 use crate::{ErrorInfo, Result};
 
-/// A 128 bit AES key.
-#[derive(Clone)]
-pub struct Key128(GenericArray<u8, U16>);
+type Aes128CbcEnc = cbc::Encryptor<aes::Aes128>;
+type Aes256CbcEnc = cbc::Encryptor<aes::Aes256>;
+type Aes128CbcDec = cbc::Decryptor<aes::Aes128>;
+type Aes256CbcDec = cbc::Decryptor<aes::Aes256>;
 
-/// A 256 bit AES key.
-#[derive(Clone)]
-pub struct Key256(GenericArray<u8, U32>);
+pub(crate) type IV = [u8; 16];
 
-// Type alias for CBC mode with Pkcs7 padding.
-type Cbc<T> = block_modes::Cbc<T, Pkcs7>;
-
-// An internal type for providing a deterministic IV in tests.
-pub(crate) type IV = [u8; aes::BLOCK_SIZE];
-
-#[derive(Clone)]
-pub enum Key {
+#[derive(Clone, Debug)]
+pub enum Cipher {
     /// A 128 bit AES key.
-    Key128(Key128),
+    Aes128Cbc([u8; 16]),
     /// A 256 bit AES key.
-    Key256(Key256),
+    Aes256Cbc([u8; 32]),
+}
+
+impl Default for Cipher {
+    fn default() -> Self {
+        Self::builder().build().unwrap()
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum CipherKind {
+    AesCbc,
+}
+
+impl Default for CipherKind {
+    fn default() -> Self {
+        CipherKind::AesCbc
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum KeyLen {
+    Bits128,
+    Bits256,
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct CipherBuilder {
+    kind: CipherKind,
+    len: Option<KeyLen>,
+    key: Option<Vec<u8>>,
+}
+
+impl CipherBuilder {
+    pub fn kind(mut self, kind: CipherKind) -> Self {
+        self.kind = kind;
+        self
+    }
+
+    pub fn key(mut self, key: Vec<u8>) -> Self {
+        self.key = Some(key);
+        self
+    }
+
+    pub fn string(mut self, key: &str) -> Result<Self> {
+        let key = base64::decode(key)?;
+        self.key = Some(key);
+        Ok(self)
+    }
+
+    pub fn len(mut self, len: KeyLen) -> Self {
+        self.len = Some(len);
+        self
+    }
+
+    pub fn build(self) -> Result<Cipher> {
+        let len = self.len.or_else(|| {
+            self.key.as_ref().and_then(|key| match key.len() {
+                16 => Some(KeyLen::Bits128),
+                32 => Some(KeyLen::Bits256),
+                _ => None,
+            })
+        });
+
+        let cipher = match self.kind {
+            CipherKind::AesCbc => match len {
+                Some(KeyLen::Bits128) => {
+                    let key = if let Some(key) = self.key {
+                        key.try_into()
+                            .map_err(|_| error!(40000, "Invalid key size"))?
+                    } else {
+                        let mut data = [0; 16];
+                        thread_rng().fill_bytes(&mut data);
```

Comment:
I do think the generation of the key should be accessible to the user, since a typical flow will be:

- generate a key
- store it somewhere
- initialise a channel with the key to encrypt messages
- load the key on another client
- initialise a channel with the key to decrypt messages

Reply:
This is already the case. Just the key and cipher are one.

Reply:
Can you give an example of how would you implement my workflow above with the code the way it is?

Reply:
So the key is generated unless set, you can pass your own key in easily. You can also get the generated key out if you wish, there could probably be a convenience function for this.

Reply:
````rust
let cipher = ChipherBuilder::new().len(KeyLen::Bits128).build()?;
let key = cipher.key();

let cipher2 = ChipherBuilder::new().key(key).build()?;
```

Reply:
Ok thanks, I think it's worth putting that [in the example](https://github.com/ably/ably-rust/blob/idiomaticify/examples/encrypt.rs#L13-L15) to make it clear to users who want to use encryption how they can retrieve the key, especially since this API will not match any of our main documentation which will refer to calling `generate_random_key`.

Reply:
Will do. We could also have a generate_random_key function that's shorthand for this.

Reply:
Do you intend to add to the example in this PR?